java/code/src/com/redhat/rhn/taskomatic/task/HubReportDbUpdateDriver.java

```diff
@@ -40,54 +39,99 @@
  */
 public class HubReportDbUpdateDriver implements QueueDriver<MgrServerInfo> {
 
+    private static Set<MgrServerInfo> currentMgrServerInfos = Collections.synchronizedSet(new HashSet<>());
     private Logger log;
-    private static AtomicBoolean hasCandidates = new AtomicBoolean(true);
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public void setLogger(Logger loggerIn) {
         this.log = loggerIn;
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public Logger getLogger() {
         return log;
     }
 
-    @Override
-    public List<MgrServerInfo> getCandidates() {
+    public static synchronized Set<MgrServerInfo> getCurrentMgrServerInfos() {
```

Comment:
As already discussed directly, this synchronisation is happening on the class, while the block in `getCandidates` is on the collection itself. This means they are not blocking each other. I'm not saying that this is an issue, but it's worth double checking this is the expected behaviour.